### PR TITLE
Add transcoding of api sample terrain_tessellation based on vulkan.hpp

### PR DESCRIPTION
**Pull request by asuessenbach**
## Description

Introduces a new Vulkan-Hpp-based sample, which is a transcoded version of the terrain_tessellation sample.
Additional adjustments:
+ added `#pragma once` to geometry/frustum.h
+ removed some unused variables from original terrain_tessellation.cppp

Build tested on Windows with NVIDIA GPU.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: samples/api/hpp_terrain_tessellation/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
```
Comment (gpx1000) [inline samples/api/hpp_terrain_tessellation/CMakeLists.txt:1]: The copyright notices in this sample are confusing me.  Sascha is listed as the author, NVIDIA is sometimes copyright holder, and sometimes Khronos Group.  I don't know who the right holder is but wanted to raise this.

File: samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+	}
+
+	HPPApiVulkanSample::submit_frame();
+}
+
+// Generate a terrain quad patch for feeding to the tessellation control shader
+void HPPTerrainTessellation::generate_terrain()
+{
+#define PATCH_SIZE 64
+#define UV_SCALE 1.0f
```
Comment (gpx1000) [inline samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp]: Minor nit here.  For readability if we must have macros, it's usually best to put them prominently at the top of the file.

File: samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+
+	get_device().get_handle().destroyBuffer(vertex_staging.buffer);
+	get_device().get_handle().freeMemory(vertex_staging.memory);
+	get_device().get_handle().destroyBuffer(index_staging.buffer);
+	get_device().get_handle().freeMemory(index_staging.memory);
+}
+
+void HPPTerrainTessellation::load_assets()
+{
+	// @todo: sascha
```
Comment (gpx1000) [inline samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp]: anyone know what the todo is meant to denote here?  I know it's in the original C version, but it raises a good point that a todo is hanging out here that should either be addressed or maybe a little more added here so we know what needs to be done.

**Review by gpx1000 — COMMENTED** [review]
minor edits requested.  Will do another pass when I get a chance.

File: samples/api/hpp_terrain_tessellation/CMakeLists.txt (hunk 1 above)
Comment (asuessenbach) [reply samples/api/hpp_terrain_tessellation/CMakeLists.txt:1]: New files introduced by me are copyrighted by NVIDIA (besides README.md... can change that to NVIDIA as well though).
Existing files that are just slightly modified just got a new year, but I did not change the copyright itself.
Actually, I don't really know what's the right (that is: legally correct) way to mark that. Would be happy to "make it right".

File: samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp (hunk 2 above)
Comment (asuessenbach) [reply samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp]: Changed that to local const variables.
Changed that as well in the original terrain_tessellation.cpp.

File: samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp (hunk 3 above)
Comment (SaschaWillems) [reply samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp]: Good question. I don't remember putting it in there, and I don't see what it should refer to. So yes, we should remove it in both versions of the sample.

File: samples/api/hpp_terrain_tessellation/CMakeLists.txt (hunk 1 above)
Comment (SaschaWillems) [reply samples/api/hpp_terrain_tessellation/CMakeLists.txt:1]: That's actually a good point. I think we need a clear guideline on how to handle copyrights when reusing code from other samples/authors. Maybe @marty-johnson59 can help us with this.

**Comment by SaschaWillems** [comment]
Aside from Steven's comments code and sample LTGM. A link to the sample needs to be added to the sample listing README.md though.

File: samples/api/hpp_terrain_tessellation/CMakeLists.txt (hunk 1 above)
Comment (marty-johnson59) [reply samples/api/hpp_terrain_tessellation/CMakeLists.txt:1]: I've got a query out to Emily on this, but my understanding (and what I've done in the past) is that it's OK to leave the copyright as the original author (NVIDIA here), but update the year to 2022 if there are changes in the file this year.  I'll get back to you if I hear anything different.

Comment (marty-johnson59) [reply samples/api/hpp_terrain_tessellation/CMakeLists.txt:1]: Emily tells me they're working on copyright guidance document right now (to be posted soon), but that keeping original author is fine.  Also fine to add others who edit/modify the file - or just list "and contributors" to keep it simple..

File: samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp (hunk 3 above)
Comment (asuessenbach) [reply samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp]: Removed.

**Comment by asuessenbach** [comment]
> Aside from Steven's comments code and sample LTGM. A link to the sample needs to be added to the sample listing README.md though.

Good point!
Added.

**Review by gary-sweet — COMMENTED** [review]
As expected, this won't run for me due to lack of tessellation support.
It does report that correctly, but I then see a segmentation fault as it exits.
That doesn't happen with the other unsupported samples I've tried.

**Comment by asuessenbach** [comment]
@gary-sweet Great catch! I never tried the failure path.
To fix this, I had to change the device getter function in HPPVulkanSample, and it's usage all over the hpp-based framework and sample files.

**Comment by gary-sweet** [comment]
> @gary-sweet Great catch! I never tried the failure path. To fix this, I had to change the device getter function in HPPVulkanSample, and it's usage all over the hpp-based framework and sample files.

Hmmm. I'm now seeing all the hpp samples segfaulting on start up.

**Comment by gary-sweet** [comment]
> Hmmm. I'm now seeing all the hpp samples segfaulting on start up.

It looks like the gpus array inside Instance() is full of garbage. The crash happens in Instance::get_first_gpu() because it pulls a null gpu from the list. It's not immediately obvious how that has happened though.

Valgrind is also showing the use of unitialised data at that point:

`==2602723== Conditional jump or move depends on uninitialised value(s)
==2602723==    at 0x328DB8: vkb::Instance::get_first_gpu() (instance.cpp:443)
==2602723==    by 0x2C7028: vkb::DirectWindow::create_surface(vkb::Instance&) (direct_window.cpp:244)
==2602723==    by 0x71429C: vkb::platform::HPPWindow::create_surface(vkb::core::HPPInstance&) (hpp_window.h:42)
==2602723==    by 0x70AB45: vkb::HPPVulkanSample::prepare(vkb::platform::HPPPlatform&) (hpp_vulkan_sample.cpp:105)
==2602723==    by 0x6F8621: HPPApiVulkanSample::prepare(vkb::platform::HPPPlatform&) (hpp_api_vulkan_sample.cpp:29)
==2602723==    by 0x49E528: HPPTerrainTessellation::prepare(vkb::platform::HPPPlatform&) (hpp_terrain_tessellation.cpp:67)
==2602723==    by 0x4585BF: vkb::platform::HPPApplication::prepare(vkb::Platform&) (hpp_application.h:44)
==2602723==    by 0x203529: vkb::Platform::start_app() (platform.cpp:380)
==2602723==    by 0x202662: vkb::Platform::main_loop() (platform.cpp:132)
==2602723==    by 0x1E5ABB: main (main.cpp:54)
==2602723== 
==2602723== Conditional jump or move depends on uninitialised value(s)
==2602723==    at 0x328E1B: vkb::Instance::get_first_gpu() (instance.cpp:446)
==2602723==    by 0x2C7028: vkb::DirectWindow::create_surface(vkb::Instance&) (direct_window.cpp:244)
==2602723==    by 0x71429C: vkb::platform::HPPWindow::create_surface(vkb::core::HPPInstance&) (hpp_window.h:42)
==2602723==    by 0x70AB45: vkb::HPPVulkanSample::prepare(vkb::platform::HPPPlatform&) (hpp_vulkan_sample.cpp:105)
==2602723==    by 0x6F8621: HPPApiVulkanSample::prepare(vkb::platform::HPPPlatform&) (hpp_api_vulkan_sample.cpp:29)
==2602723==    by 0x49E528: HPPTerrainTessellation::prepare(vkb::platform::HPPPlatform&) (hpp_terrain_tessellation.cpp:67)
==2602723==    by 0x4585BF: vkb::platform::HPPApplication::prepare(vkb::Platform&) (hpp_application.h:44)
==2602723==    by 0x203529: vkb::Platform::start_app() (platform.cpp:380)
==2602723==    by 0x202662: vkb::Platform::main_loop() (platform.cpp:132)
==2602723==    by 0x1E5ABB: main (main.cpp:54)
==2602723== 
==2602723== Invalid read of size 8`

**Comment by asuessenbach** [comment]
@gary-sweet That's strange! I don't see anything like that.
Could you please verify that you get something meaningful in `HPPInstance::query_gpus()` (core/hpp_instance.cpp, line 441)?

**Comment by gary-sweet** [comment]
> @gary-sweet That's strange! I don't see anything like that. Could you please verify that you get something meaningful in `HPPInstance::query_gpus()` (core/hpp_instance.cpp, line 441)?

Yes, I can see a single gpu being added at that point. The gpus array looks good there. By the time it gets out of "instance = std::make_unique<vkb::core::HPPInstance>(get_name(), get_instance_extensions(), get_validation_layers(), headless, api_version);" on line 102 of hpp_vulkan_sample.cpp however, the gpus array in instance looks like it contains garbage.

**Comment by SaschaWillems** [comment]
This sounds like a compiler-specific issue. I don't see that problem on Windows with MSVC, but Android uses a different compiler.

**Comment by gary-sweet** [comment]
> This sounds like a compiler-specific issue. I don't see that problem on Windows with MSVC, but Android uses a different compiler.

I get the same issue with two different versions of gcc in Linux

**Comment by gary-sweet** [comment]
It might be something related to using DirectWindow perhaps?

**Comment by asuessenbach** [comment]
That is, at the end of the HPPInstance constructor (hpp_instance.cpp, line 346), the gpus array looks fine, but the shared_ptr created from that (hpp_vulkan_sample.cpp, line 102) looks bad?
That would imply that the make_unique did something, well, unexpected. Doesn't sound like a DirectWindow specific issue.
But I don't see how that could happen.

Is it "just" the gpus array, that look like garbage? The other data in the HPPInstance look reasonable?

**Comment by asuessenbach** [comment]
@gary-sweet Any new findings on your issue?

**Comment by gary-sweet** [comment]
> @gary-sweet Any new findings on your issue?

Sorry, not yet. I've not had a lot of time to spend on it. I'm hoping to find a little time before today's meeting, but I can't guarantee anything I'm afraid.

**Comment by gary-sweet** [comment]
> > @gary-sweet Any new findings on your issue?
> 
> Sorry, not yet. I've not had a lot of time to spend on it. I'm hoping to find a little time before today's meeting, but I can't guarantee anything I'm afraid.

Ok - I think I've found the culprit. 

There is a reinterpret_cast from `vkb::core::HPPInstance` in `HPPWindow::create_surface()` that looks like this:
`return static_cast<vk::SurfaceKHR>(create_surface(reinterpret_cast<vkb::Instance &>(instance)));`

As far as I can see, `vkb::Instance` and `vkb::core::HPPInstance` are unrelated classes, so the cast results in a garbage object.

Changing `create_surface()` instead to do this fixes the issues for me:
`return static_cast<vk::SurfaceKHR>(create_surface(instance.get_handle(), instance.get_first_gpu().get_handle()));`

**Comment by asuessenbach** [comment]
@gary-sweet That's a bit surprising, as the two Instance flavours are supposed to be bit-identical.
But so what, great that you've figured out the actual culprit and already found a fix! Thanks a lot!

**Comment by gary-sweet** [comment]
> @gary-sweet That's a bit surprising, as the two Instance flavours are supposed to be bit-identical. But so what, great that you've figured out the actual culprit and already found a fix! Thanks a lot!

Thanks. Now correctly reports that the test is unsupported without crashing. Not sure I can really +1 it on that basis though.

**Comment by gary-sweet** [comment]
> .... the two Instance flavours are supposed to be bit-identical

That sounds like an accident waiting to happen to me (and one that seems to have already happened). Is there no better way to structure the classes that enforces that relationship?

**Comment by asuessenbach** [comment]
@gary-sweet Unfortunately, I don't see any other (let alone, better) way to structures those classes.
But in order to possibly get some more insights into this issue, would you please replace your approach
`return static_cast<vk::SurfaceKHR>(create_surface(instance.get_handle(), instance.get_first_gpu().get_handle()));`
with this one:
`return static_cast<vk::SurfaceKHR>(create_surface(*static_cast<vkb::Instance *>(static_cast<void *>(&instance))));`

Moreover, if you could temporarily make the members of the classes vkb::core::HPPInstance, vkb::core::HPPPhysicalDevice, vkb::Instance, and vkb::PhysicalDevice public, and add this little sequence of static assertions to your code, to see if any of them fires?
```
static_assert(sizeof(vkb::core::HPPInstance) == sizeof(vkb::Instance), "wrong size");
static_assert(offsetof(vkb::core::HPPInstance, handle) == offsetof(vkb::Instance, handle), "wrong offset of handle");
static_assert(sizeof(vkb::core::HPPInstance::handle) == sizeof(vkb::Instance::handle), "wrong size of handle");
static_assert(offsetof(vkb::core::HPPInstance, enabled_extensions) == offsetof(vkb::Instance, enabled_extensions), "wrong offset of enabled_extensions");
static_assert(sizeof(vkb::core::HPPInstance::enabled_extensions) == sizeof(vkb::Instance::enabled_extensions), "wrong size of enabled_extensions");
static_assert(offsetof(vkb::core::HPPInstance, debug_utils_messenger) == offsetof(vkb::Instance, debug_utils_messenger), "wrong offset of debug_utils_messenger");
static_assert(sizeof(vkb::core::HPPInstance::debug_utils_messenger) == sizeof(vkb::Instance::debug_utils_messenger), "wrong size of debug_utils_messenger");
static_assert(offsetof(vkb::core::HPPInstance, debug_report_callback) == offsetof(vkb::Instance, debug_report_callback), "wrong offset of debug_report_callback");
static_assert(sizeof(vkb::core::HPPInstance::debug_report_callback) == sizeof(vkb::Instance::debug_report_callback), "wrong size of debug_report_callback");
static_assert(offsetof(vkb::core::HPPInstance, gpus) == offsetof(vkb::Instance, gpus), "wrong offset of gpus");
static_assert(sizeof(vkb::core::HPPInstance::gpus) == sizeof(vkb::Instance::gpus), "wrong size of gpus");

static_assert(sizeof(vkb::core::HPPPhysicalDevice) == sizeof(vkb::PhysicalDevice), "wrong size");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, instance) == offsetof(vkb::PhysicalDevice, instance), "wrong offset of instance");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::instance) == sizeof(vkb::PhysicalDevice::instance), "wrong size of instance");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, handle) == offsetof(vkb::PhysicalDevice, handle), "wrong offset of handle");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::handle) == sizeof(vkb::PhysicalDevice::handle), "wrong size of handle");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, features) == offsetof(vkb::PhysicalDevice, features), "wrong offset of features");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::features) == sizeof(vkb::PhysicalDevice::features), "wrong size of features");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, properties) == offsetof(vkb::PhysicalDevice, properties), "wrong offset of properties");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::properties) == sizeof(vkb::PhysicalDevice::properties), "wrong size of properties");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, memory_properties) == offsetof(vkb::PhysicalDevice, memory_properties), "wrong offset of memory_properties");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::memory_properties) == sizeof(vkb::PhysicalDevice::memory_properties), "wrong size of memory_properties");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, queue_family_properties) == offsetof(vkb::PhysicalDevice, queue_family_properties), "wrong offset of queue_family_properties");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::queue_family_properties) == sizeof(vkb::PhysicalDevice::queue_family_properties), "wrong size of queue_family_properties");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, requested_features) == offsetof(vkb::PhysicalDevice, requested_features), "wrong offset of requested_features");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::requested_features) == sizeof(vkb::PhysicalDevice::requested_features), "wrong size of requested_features");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, last_requested_extension_feature) == offsetof(vkb::PhysicalDevice, last_requested_extension_feature), "wrong offset of last_requested_extension_feature");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::last_requested_extension_feature) == sizeof(vkb::PhysicalDevice::last_requested_extension_feature), "wrong size of last_requested_extension_feature");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, extension_features) == offsetof(vkb::PhysicalDevice, extension_features), "wrong offset of extension_features");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::extension_features) == sizeof(vkb::PhysicalDevice::extension_features), "wrong size of extension_features");
static_assert(offsetof(vkb::core::HPPPhysicalDevice, high_priority_graphics_queue) == offsetof(vkb::PhysicalDevice, high_priority_graphics_queue), "wrong offset of high_priority_graphics_queue");
static_assert(sizeof(vkb::core::HPPPhysicalDevice::high_priority_graphics_queue) == sizeof(vkb::PhysicalDevice::high_priority_graphics_queue), "wrong size of high_priority_graphics_queue");
```

**Comment by gary-sweet** [comment]
> replace your approach `return static_cast<vk::SurfaceKHR>(create_surface(instance.get_handle(), instance.get_first_gpu().get_handle()));` with this one: `return static_cast<vk::SurfaceKHR>(create_surface(*static_cast<vkb::Instance *>(static_cast<void *>(&instance))));`

Starts crashing again I'm afraid.


> Moreover, if you could temporarily make the members of the classes vkb::core::HPPInstance, vkb::core::HPPPhysicalDevice, vkb::Instance, and vkb::PhysicalDevice public, and add this little sequence of static assertions to your code, to see if any of them fires?

They don't get as far as firing, I see a whole bunch of compiler errors like this:

```
error: offsetof within non-standard-layout type ‘vkb::core::HPPPhysicalDevice’ is conditionally-supported [-Werror=invalid-offsetof]
   98 | static_assert(offsetof(vkb::core::HPPPhysicalDevice, instance) == offsetof(vkb::PhysicalDevice, instance), "wrong offset of instance");
```

If I disable that warning, none of the static asserts fire.

**Comment by gary-sweet** [comment]
In a debugger it looks like the gpus array inside instance is fine inside the constructor of the instance, but immediately after that, in the calling code, the gpus array is corrupt.
`instance = std::make_unique<vkb::core::HPPInstance>(get_name(), get_instance_extensions(), get_validation_layers(), headless, api_version);`
instance->gpus is bad now

(In fact if I `printf("instance->gpus.size() = %zu\n", instance->gpus.size());` here it reports `instance->gpus.size() = 18446744073709549517`)

**Comment by asuessenbach** [comment]
That is, it's not the cast from vkb::core::HPPInstance to vbk::Instance that fails; the vkb::core::HPPInstance passed into create_surface is already corrupted??
That issue turns from strange to mysterious!

What do you get with an instruction like that:
`vkb::core::HPPInstance *inst = new vkb::core::HPPInstance(get_name(), get_instance_extensions(), get_validation_layers(), headless, api_version);
`
Is the `inst` valid then?

And what do you get with 
`std::shared_ptr<vkb::core::HPPInstance> sharedInst = std::make_shared<vkb::core::HPPInstance>(get_name(), get_instance_extensions(), get_validation_layers(), headless, api_version);
`

And finally, what do you get with
`instance.reset(new vkb::core::HPPInstance(get_name(), get_instance_extensions(), get_validation_layers(), headless, api_version) );
`

Just to be sure, the corresponding instruction in vulkan_sample.cpp, line 129, (with `sample terrain_tessellation`, for example) gets you a valid instance?

**Comment by gary-sweet** [comment]
> That issue turns from strange to mysterious!

100% agree!

> What do you get with an instruction like that: `vkb::core::HPPInstance *inst = new vkb::core::HPPInstance(get_name(), get_instance_extensions(), get_validation_layers(), headless, api_version); ` Is the `inst` valid then?

No. Same result.

> And what do you get with `std::shared_ptr<vkb::core::HPPInstance> sharedInst = std::make_shared<vkb::core::HPPInstance>(get_name(), get_instance_extensions(), get_validation_layers(), headless, api_version); `

Same result.

> And finally, what do you get with `instance.reset(new vkb::core::HPPInstance(get_name(), get_instance_extensions(), get_validation_layers(), headless, api_version) ); `

Same result.

> Just to be sure, the corresponding instruction in vulkan_sample.cpp, line 129, (with `sample terrain_tessellation`, for example) gets you a valid instance?

It does.

I'm going to investigate what happens to the gpus array. Maybe a data watchpoint will help figure out what's going on.

**Comment by gary-sweet** [comment]
So, simply printing the address of the gpus vector inside the constructor and outside show that each thinks it's in a slightly different place.
```
&gpus = 0x5644c1178c10
&gpus = 0x5644c1178c20

```

**Comment by gary-sweet** [comment]
Right - so the issue is related to the `#if defined(VKB_DEBUG) || defined(VKB_VALIDATION_LAYERS)` section in `HPPInstance`.
If I comment out the `#if` line, so those fields are always there, the problems go away.

So, somehow, different files are seeing different values for either VKB_DEBUG or VKB_VALIDATION_LAYERS.

**Comment by gary-sweet** [comment]
Annoyingly, that's something I did consider very early on but couldn't see how that could be the case, so dismissed it.

**Comment by gary-sweet** [comment]
Looks like many of the hpp_ variant files aren't including common/helpers.h or common/error.h which is what defines VKB_DEBUG  afaict. That will need addressing in some way. It might be safer to just always have those members in the classes tbh, they don't seem to be large.

**Comment by asuessenbach** [comment]
Elementary!

I see, VKB_DEBUG is defined in common/error.h, so I need to include that in (most of) the hpp-headers. But where's `VKB_VALIDATION_LAYERS` and `VKB_VALIDATION_LAYERS_GPU_ASSISTED` defined? I can't find them.

**Comment by gary-sweet** [comment]
> Where's `VKB_VALIDATION_LAYERS` and `VKB_VALIDATION_LAYERS_GPU_ASSISTED` defined? I can't find them.

I can only see a CMake variable being set. I don't see that being converted into a compile-time definition though.

**Comment by gary-sweet** [comment]
Looks much better with the latest change. Thanks

**Comment by asuessenbach** [comment]
@gary-sweet Thanks to you for running this sample using such a nitpicking compiler! MSVC is much more generous, in that respect :(

**Review by gpx1000 — APPROVED** [review]
LGTM;

Excellent job on finding that problem and fixing it was fun to watch.

**Review by SaschaWillems — APPROVED** [review]
LGTM and works fine for me on Windows with an RTX 2060 using MSVC.

**Pull request merged**